src/ring_theory/free_comm_ring.lean

```diff
@@ -0,0 +1,267 @@
+
+end free_comm_ring
+
+def free_comm_ring_pempty_equiv_int : free_comm_ring pempty.{u+1} ≃r ℤ :=
+{ to_fun := free_comm_ring.lift $ pempty.rec _,
+  inv_fun := coe,
+  left_inv := λ x, free_abelian_group.induction_on x rfl
+    (λ s, multiset.induction_on s rfl $ pempty.rec _)
+    (λ s ih, by rw [free_comm_ring.lift_neg, int.cast_neg, ih])
+    (λ x1 x2 ih1 ih2, by rw [free_comm_ring.lift_add, int.cast_add, ih1, ih2]),
+  right_inv := λ i, int.induction_on i rfl
+    (λ i ih, by rw [int.cast_add, int.cast_one, free_comm_ring.lift_add, free_comm_ring.lift_one, ih])
+    (λ i ih, by rw [int.cast_sub, int.cast_one, free_comm_ring.lift_sub, free_comm_ring.lift_one, ih]),
+  hom := by apply_instance }
+
+def free_comm_ring_punit_equiv_polynomial_int : free_comm_ring punit.{u+1} ≃r polynomial ℤ :=
+{ to_fun := free_comm_ring.lift $ λ _, polynomial.X,
+  inv_fun := polynomial.eval₂ coe (free_comm_ring.of punit.star),
+  left_inv := λ x, begin
+    haveI : is_semiring_hom (coe : int → free_comm_ring punit.{u+1}) :=
+      @@is_ring_hom.is_semiring_hom _ _ _ (@@int.cast.is_ring_hom _),
+    exact free_abelian_group.induction_on x rfl
+    (λ s, multiset.induction_on s rfl $ λ hd tl ih, show polynomial.eval₂ coe (free_comm_ring.of punit.star)
+          (free_comm_ring.lift (λ (_x : punit), polynomial.X) (free_comm_ring.of hd * free_abelian_group.of tl)) =
+        free_comm_ring.of hd * free_abelian_group.of tl,
+      by cases hd; rw [free_comm_ring.lift_mul, free_comm_ring.lift_of, polynomial.eval₂_mul, polynomial.eval₂_X, ih])
+    (λ s ih, by rw [free_comm_ring.lift_neg, ← neg_one_mul, polynomial.eval₂_mul,
+        ← polynomial.C_1, ← polynomial.C_neg, polynomial.eval₂_C,
+        int.cast_neg, int.cast_one, neg_one_mul, ih])
+    (λ x1 x2 ih1 ih2, by rw [free_comm_ring.lift_add, polynomial.eval₂_add, ih1, ih2])
+  end,
+  right_inv := λ x, begin
+    haveI : is_semiring_hom (coe : int → free_comm_ring punit.{u+1}) :=
+      @@is_ring_hom.is_semiring_hom _ _ _ (@@int.cast.is_ring_hom _),
+    have : ∀ i : ℤ, free_comm_ring.lift (λ _ : punit, polynomial.X) ↑i = polynomial.C i,
+    { exact λ i, int.induction_on i
+      (by rw [int.cast_zero, free_comm_ring.lift_zero, polynomial.C_0])
+      (λ i ih, by rw [int.cast_add, int.cast_one, free_comm_ring.lift_add, free_comm_ring.lift_one, ih, polynomial.C_add, polynomial.C_1])
+      (λ i ih, by rw [int.cast_sub, int.cast_one, free_comm_ring.lift_sub, free_comm_ring.lift_one, ih, polynomial.C_sub, polynomial.C_1]) },
+    exact polynomial.induction_on x
+    (λ i, by rw [polynomial.eval₂_C, this])
+    (λ p q ihp ihq, by rw [polynomial.eval₂_add, free_comm_ring.lift_add, ihp, ihq])
+    (λ n i ih, by rw [polynomial.eval₂_mul, polynomial.eval₂_pow, polynomial.eval₂_C, polynomial.eval₂_X,
+        free_comm_ring.lift_mul, free_comm_ring.lift_pow, free_comm_ring.lift_of, this])
+  end,
+  hom := by apply_instance }
+
+variables (α : Type u) [comm_ring α]
+
+def polynomial' : Type u :=
```

Comment:
This seems like an awfully complicated and unreadable definition. Can't it be split into several subdefinitions?

Reply:
I think this can and should be generalized to the multivariate case. The definition itself can be written as `(ideal.span {x | rel x}).quotient` where `rel` is an inductive predicate on A + B that says that the elements of B (which is a ring) are related to the equivalent ring expressions.

Reply:
I'll probably delete this if doing so increases the chance of getting it merged.

Reply:
deleted